### ISSUE TITLE
Mix-FFN in SegFormer

### ISSUE THREAD

**Issue by xaclincoln** [open-mmlab/mmsegmentation#1287]
quoted from Segformer 3.1 Mix-FFN part

> ViT uses positional encoding (PE) to introduce the location information. However, the
> resolution of PE is fixed. Therefore, when the test resolution is different from the training one, the
> positional code needs to be interpolated and this often leads to dropped accuracy. To alleviate this
> problem, CPVT [54] uses 3 × 3 Conv together with the PE to implement a data-driven PE. We argue
> that positional encoding is actually not necessary for semantic segmentation. Instead, we introduce
> Mix-FFN which considers the effect of zero padding to leak location information [69], by directly
> using a 3 × 3 Conv in the feed-forward network (FFN). Mix-FFN can be formulated as:
> xout = MLP(GELU(Conv3×3(MLP(xin)))) + xin, (3)
> where xin is the feature from the self-attention module. Mix-FFN mixes a 3 × 3 convolution and
> an MLP into each FFN. In our experiments, we will show that a 3 × 3 convolution is sufficient to
> provide positional information for Transformers. In particular, we use depth-wise convolutions for
> reducing the number of parameters and improving efficiency

[Here in code](https://github.com/open-mmlab/mmsegmentation/blob/c23e90289616c4797e7f329e271ca1ae4740c751/mmseg/models/backbones/mit.py), 1 by 1 convolution is used instead of MLP, is this by design or a mistake?
```
class MixFFN(BaseModule):
    """An implementation of MixFFN of Segformer.

    The differences between MixFFN & FFN:
        1. Use 1X1 Conv to replace Linear layer.
        2. Introduce 3X3 Conv to encode positional information.
    Args:
        embed_dims (int): The feature dimension. Same as
            `MultiheadAttention`. Defaults: 256.
        feedforward_channels (int): The hidden dimension of FFNs.
            Defaults: 1024.
        act_cfg (dict, optional): The activation config for FFNs.
            Default: dict(type='ReLU')
        ffn_drop (float, optional): Probability of an element to be
            zeroed in FFN. Default 0.0.
        dropout_layer (obj:`ConfigDict`): The dropout_layer used
            when adding the shortcut.
        init_cfg (obj:`mmcv.ConfigDict`): The Config for initialization.
            Default: None.
    """

    def __init__(self,
                 embed_dims,
                 feedforward_channels,
                 act_cfg=dict(type='GELU'),
                 ffn_drop=0.,
                 dropout_layer=None,
                 init_cfg=None):
        super(MixFFN, self).__init__(init_cfg)

        self.embed_dims = embed_dims
        self.feedforward_channels = feedforward_channels
        self.act_cfg = act_cfg
        self.activate = build_activation_layer(act_cfg)

        in_channels = embed_dims
        fc1 = Conv2d(  **#check here**
            in_channels=in_channels,
            out_channels=feedforward_channels,
            kernel_size=1,
            stride=1,
            bias=True)
        # 3x3 depth wise conv to provide positional encode information
        pe_conv = Conv2d(
            in_channels=feedforward_channels,
            out_channels=feedforward_channels,
            kernel_size=3,
            stride=1,
            padding=(3 - 1) // 2,
            bias=True,
            groups=feedforward_channels)
        fc2 = Conv2d(
            in_channels=feedforward_channels,
            out_channels=in_channels,
            kernel_size=1,
            stride=1,
            bias=True)
        drop = nn.Dropout(ffn_drop)
        layers = [fc1, pe_conv, self.activate, drop, fc2, drop]
        self.layers = Sequential(*layers)
        self.dropout_layer = build_dropout(
            dropout_layer) if dropout_layer else torch.nn.Identity()

    def forward(self, x, hw_shape, identity=None):
        out = nlc_to_nchw(x, hw_shape)
        out = self.layers(out)
        out = nchw_to_nlc(out)
        if identity is None:
            identity = x
        return identity + self.dropout_layer(out)

```
Crossed check the implementation in transformer repository, linear layer is used accordingly. 
In my view, linear layer would cause model to see more global context info.
Following is the code from the transformer SegFormer implementation:

```
class SegformerMixFFN(nn.Module):
    def __init__(self, config, in_features, hidden_features=None, out_features=None):
        super().__init__()
        out_features = out_features or in_features
        self.dense1 = nn.Linear(in_features, hidden_features) **#check here**
        self.dwconv = SegformerDWConv(hidden_features)
        if isinstance(config.hidden_act, str):
            self.intermediate_act_fn = ACT2FN[config.hidden_act]
        else:
            self.intermediate_act_fn = config.hidden_act
        self.dense2 = nn.Linear(hidden_features, out_features)
        self.dropout = nn.Dropout(config.hidden_dropout_prob)

    def forward(self, hidden_states, height, width):
        hidden_states = self.dense1(hidden_states)
        hidden_states = self.dwconv(hidden_states, height, width)
        hidden_states = self.intermediate_act_fn(hidden_states)
        hidden_states = self.dropout(hidden_states)
        hidden_states = self.dense2(hidden_states)
        hidden_states = self.dropout(hidden_states)
        return hidden_states
```

**Comment by RockeyCoss** [comment]
Well

**Comment by RockeyCoss** [comment]
For example, we have a tensor `input` with shape [N, C, H, W].

If we use linear layer to transform it, we need to reshape it to [N, C, HxW] and then transpose the tensor to [N, HxW, C]. Therefore, the linear layer is actually applied on `C` channel.

And if we use 1x1 convolution layer to transform it, the convolution is also applied on `C` channel. 1x1 convolution and linear layer is equivalent.

The advantage of using convolution is that it reduces the number of reshape process.(Because we need to reshape the [N, HxW, C] tensor to shape [N, C, H, W] before adding positional information)

**Comment by xaclincoln** [comment]
Yes, they are almost equivalent. But would bias in linear layer contribute since there are more parameters than 1by1 conv?
Checked the official implementation, some linear layers are replaced by 1*1 conv indeed.

**Comment by RockeyCoss** [comment]
> 

You can set `bias=True` in the convolution layer.
